src/libData/BlockChainData/BlockChain.h

```diff
@@ -0,0 +1,153 @@
+/**
+* Copyright (c) 2018 Zilliqa 
+* This source code is being disclosed to you solely for the purpose of your participation in 
+* testing Zilliqa. You may view, compile and run the code for that purpose and pursuant to 
+* the protocols and algorithms that are programmed into, and intended by, the code. You may 
+* not do anything else with the code without express permission from Zilliqa Research Pte. Ltd., 
+* including modifying or publishing the code (or any part of it), and developing or forming 
+* another public or private blockchain network. This source code is provided ‘as is’ and no 
+* warranties are given as to title or non-infringement, merchantability or fitness for purpose 
+* and, to the extent permitted by law, all liability for your use of the code is disclaimed. 
+* Some programs in this code are governed by the GNU General Public License v3.0 (available at 
+* https://www.gnu.org/licenses/gpl-3.0.en.html) (‘GPLv3’). The programs that are governed by 
+* GPLv3.0 are those programs that are located in the folders src/depends and tests/depends 
+* and which include a reference to GPLv3 in their program files.
+**/
+
+#ifndef __BLOCKCHAIN_H__
+#define __BLOCKCHAIN_H__
+
+#include <mutex>
+#include <vector>
+
+#include <boost/multiprecision/cpp_int.hpp>
+
+#include "libData/BlockData/Block/DSBlock.h"
+#include "libData/DataStructures/CircularArray.h"
+#include "libPersistence/BlockStorage.h"
+
+/// Transient storage for DS blocks.
+template<class T> class BlockChain
+{
+    std::mutex m_mutexBlocks;
+    CircularArray<T> m_blocks;
+
+public:
+    /// Constructor.
+    BlockChain()
+    {
+        m_blocks.resize(is_same<T, DSBlock>::value ? DS_BLOCKCHAIN_SIZE
+                                                   : TX_BLOCKCHAIN_SIZE);
+    }
+
+    /// Destructor.
+    ~BlockChain() {}
+
+    /// Reset
+    void Reset()
+    {
+        m_blocks.resize(is_same<T, DSBlock>::value ? DS_BLOCKCHAIN_SIZE
+                                                   : TX_BLOCKCHAIN_SIZE);
+    }
+
+    /// Returns the number of blocks.
+    boost::multiprecision::uint256_t GetBlockCount()
```

Comment:
I can see this used like this:

```
 m_mediator.m_currentEpochNum
        = (uint64_t)m_mediator.m_txBlockChain.GetBlockCount();
```

It's impossible to use this API correctly in a threaded environment (unless we have further context), because by the time `GetBlockCount()` returns, someone else might have increased the number of blocks.